ports/samd/samd_qspiflash.c

```diff
@@ -0,0 +1,492 @@
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "py/mperrno.h"
+#include "modmachine.h"
+#include "extmod/machine_spi.h"
+#include "extmod/vfs.h"
+#include "pin_af.h"
+#include "clock_config.h"
+#include "sam.h"
+
+#ifdef MICROPY_HW_QSPIFLASH
+
+#include "drivers/memory/external_flash_device.h"
+
+// QSPI command codes
+enum
+{
+    QSPI_CMD_READ              = 0x03,
+    QSPI_CMD_READ_4B           = 0x13,
+    QSPI_CMD_QUAD_READ         = 0x6B,// 1 line address, 4 line data
+
+    QSPI_CMD_READ_JEDEC_ID     = 0x9f,
+
+    QSPI_CMD_PAGE_PROGRAM      = 0x02,
+    QSPI_CMD_PAGE_PROGRAM_4B   = 0x12,
+    QSPI_CMD_QUAD_PAGE_PROGRAM = 0x32, // 1 line address, 4 line data
+
+    QSPI_CMD_READ_STATUS       = 0x05,
+    QSPI_CMD_READ_STATUS2      = 0x35,
+
+    QSPI_CMD_WRITE_STATUS      = 0x01,
+    QSPI_CMD_WRITE_STATUS2     = 0x31,
+
+    QSPI_CMD_ENABLE_RESET      = 0x66,
+    QSPI_CMD_RESET             = 0x99,
+
+    QSPI_CMD_WRITE_ENABLE      = 0x06,
+    QSPI_CMD_WRITE_DISABLE     = 0x04,
+
+    QSPI_CMD_ERASE_SECTOR      = 0x20,
+    QSPI_CMD_ERASE_SECTOR_4B   = 0x21,
+    QSPI_CMD_ERASE_BLOCK       = 0xD8,
+    QSPI_CMD_ERASE_CHIP        = 0xC7,
+
+    QSPI_CMD_READ_SFDP_PARAMETER = 0x5A,
+};
+
+// QSPI flash pins are: CS=PB11, SCK=PB10, IO0-IO3=PA08, PA09, PA10 and PA11.
+#define PIN_CS  (43)
+#define PIN_SCK (42)
+#define PIN_IO0 (8)
+#define PIN_IO1 (9)
+#define PIN_IO2 (10)
+#define PIN_IO3 (11)
+
+#define PAGE_SIZE (256)
+#define SECTOR_SIZE (4096)
+
+typedef struct _samd_qspiflash_obj_t {
+    mp_obj_base_t base;
+    uint16_t pagesize;
+    uint16_t sectorsize;
+    uint32_t size;
+    uint8_t phase;
+    uint8_t polarity;
+} samd_qspiflash_obj_t;
+
+/// List of all possible flash devices used by Adafruit boards
+static const external_flash_device possible_devices[] = {
+    MICROPY_HW_QSPIFLASH
+};
+
+#define EXTERNAL_FLASH_DEVICE_COUNT MP_ARRAY_SIZE(possible_devices)
+static external_flash_device const *flash_device;
+static external_flash_device generic_config = GENERIC;
+extern const mp_obj_type_t samd_qspiflash_type;
+
+// The QSPIflash object is a singleton
+static samd_qspiflash_obj_t qspiflash_obj = { { &samd_qspiflash_type } };
+
+// Turn off cache and invalidate all data in it.
+static void samd_peripherals_disable_and_clear_cache(void) {
+    CMCC->CTRL.bit.CEN = 0;
+    while (CMCC->SR.bit.CSTS) {
+    }
+    CMCC->MAINT0.bit.INVALL = 1;
+}
+
+// Enable cache
+static void samd_peripherals_enable_cache(void) {
+    CMCC->CTRL.bit.CEN = 1;
+}
+
+/*
+    @brief Run a single QSPI instruction.
+    @param command instruction code
+    @param iframe iframe register value (configured by caller according to command code)
+    @param addr the address to read or write from. If the instruction doesn't require an address, this parameter is meaningless.
+    @param buffer pointer to the data to be written or stored depending on the type is Read or Write
+    @param size the number of bytes to read or write.
```

Comment:
This comment style (using `@brief`/`@param`) doesn't match anywhere else in this file or port (?).  And I couldn't find anything similar in the Adafruit code.  So just wondering where it came from, did you write it yourself?

Reply:
It came form the Adafruit MIT licensed code, which I reused. But I can delete it.